servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/PublisherFlatMapSingle.java

```diff
@@ -193,16 +192,42 @@ public void onError(Throwable t) {
 
         @Override
         public void onComplete() {
-            // active must be checked after setting the terminal event, because they are accessed in the reverse way in
-            // FlatMapSingleSubscriber and if it were reversed here the FlatMapSingleSubscriber would be racy and may
-            // not detect the terminal event.
-            if (trySetTerminal(complete(), false, terminalNotificationUpdater, this) && active == 0) {
-                // Since onComplete and onNext can not be concurrent and onNext must not be invoked post onComplete,
-                // if we see active == 0 here, active must not change after this.
+            // Setting terminal must be done before terminateActiveMappedSources to ensure visibility of the terminal.
+            final boolean setTerminal = trySetTerminal(complete(), false, terminalNotificationUpdater, this);
+            final boolean allSourcesTerminated = terminateActiveMappedSources();
+            if (setTerminal && allSourcesTerminated) {
                 enqueueAndDrain(complete());
             }
         }
 
+        private boolean terminateActiveMappedSources() {
+            for (;;) {
+                final int prevActiveMappedSources = activeMappedSources;
+                // Should always be >= 0, but just in case there is a bug in user code that results in multiple terminal
```

Comment:
Is this comment saying that `prevActiveMappedSources` should always be >= 0? We are setting it to `-prevActiveMappedSources`, so this is confusing. Maybe just elaborate more like:
```
// prevActiveMappedSources >= 0 if only a single terminal event is received (expected)
// prevActiveMappedSources < 0 if multiple terminal events are received (unexpected)
// We return true only for the first terminal event.

Reply:
good point. I'll simplify to an asset and clarify the comment.